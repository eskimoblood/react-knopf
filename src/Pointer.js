import React from 'react';


const presets = {
  circle({transform, width: r, cx, cy, style}) {
    return <circle {...{r, style, transform, cx, cy}}/>
  },

  rect({x, y, width, height, transform, style}) {
    return <rect {...{style, transform, width, height, x, y}}/>
  },

  triangle({width, height, transform, style}){
    return <path {...{style, transform}} d={`M0 0 L${width/2} ${height} L-${width/2} ${height} Z`}/>
  }
};


function getTransformation( {angle, center, radius}) {
  return {
    transform: `rotate(${angle} ${center} ${center}) translate(${center} ${center - (radius || center)}) `
  }
}

function getPosition({type, props:{r, ry, width}}) {
  switch (type) {
  case 'ellipse': {
    return {
      cx: 0,
      cy: ry
    };
  }
  case 'circle':
    return {
      cx: 0,
      cy: r
    };
  case 'rect': {
    return {
      x: -width / 2,
      y: 0
    };
  }
  default : {
    return {

    }
  }
  }
}

function limitedProps({center, knobSize, angle, value, ...rest}) {
  return rest;
}

function Pointer(props) {

  const propsWithTransformation = getTransformation(props);
  if (props.children) {
    return <g>{React.Children.map(props.children, c => {
      if (typeof c.type === 'string') {
        const propsWithPosition = getPosition(c);
        return React.cloneElement(c, Object.assign(propsWithPosition, propsWithTransformation));
      } else {
        return React.cloneElement(c, Object.assign(propsWithTransformation, props));
      }
    })}</g>
  } else {
    const {type = 'circle'} = props;
    const propsWithPosition = getPosition({type, props});
    const prop = Object.assign(propsWithPosition, propsWithTransformation, limitedProps(props));
    return presets[type](prop);
  }
}

export default Pointer;