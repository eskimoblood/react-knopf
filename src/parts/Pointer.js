import React, { PropTypes } from 'react';
import { getPosition, presets, getTransformation, limitedProps, normalizeProps } from '../utils/util';

function Pointer(props) {

  const propsWithTransformation = getTransformation(props);
  if (props.children) {
    return <g>{React.Children.map(props.children, c => {
      if (typeof c.type === 'string') {
        const propsWithPosition = getPosition(c);
        return React.cloneElement(c, Object.assign(propsWithPosition, propsWithTransformation));
      } else {
        return React.cloneElement(c, Object.assign(propsWithTransformation, props));
      }
    })}</g>
  } else {
    const { type = 'circle' } = (props);
    const normalizedProps = normalizeProps(props);
    const propsWithPosition = getPosition({ type, props: normalizedProps });
    const prop = Object.assign(normalizedProps, propsWithPosition, propsWithTransformation, limitedProps(props));
    return presets[type](prop);
  }
}

Pointer.propTypes = {
  type: PropTypes.string
};

Pointer.defaultProps = {};

export default Pointer;
