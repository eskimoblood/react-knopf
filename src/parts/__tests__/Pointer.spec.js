import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Pointer from '../Pointer';

jest.mock('', () => '');

describe('Pointer', () => {

  describe('with children', () => {

    it('of type rect', () => {
      const pointer = shallow(
        <Pointer transform="someTransformation" angle={0} center={50} radius={100}>
          <rect width={10} height={10} />
        </Pointer>
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });
    it('of type ellipse', () => {
      const pointer = shallow(
        <Pointer transform="someTransformation" angle={0} center={50} radius={100}>
          <ellipse rx={10} ry={10} />
        </Pointer>
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });

    it('of type circle', () => {
      const pointer = shallow(
        <Pointer transform="someTransformation" angle={0} center={50} radius={100}>
          <circle r={10} />
        </Pointer>
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });

    it('for custom component', () => {
      const Test = ()=> <div />
      const pointer = shallow(
        <Pointer transform="someTransformation" angle={0} center={50} radius={100}>
          <Test />
        </Pointer>
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });

  });

  describe('without children', () => {

    it('use the type prop to render the pointer', () => {
      const pointer = shallow(
        <Pointer
          type="rect"
          width={10}
          height={10}
          transform="someTransformation"
        />
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });

    it('without type prop renders a circle by default', () => {
      const pointer = shallow(
        <Pointer
          width={10}
          transform="someTransformation"
        />
      );
      expect(shallowToJson(pointer)).toMatchSnapshot();
    });

  });

});



