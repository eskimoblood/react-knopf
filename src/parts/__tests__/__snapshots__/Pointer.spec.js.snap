exports[`Pointer with children for custom component 1`] = `
<g>
  <Test
    angle={0}
    center={50}
    radius={100}
    transform="someTransformation">
    <Test />
  </Test>
</g>
`;

exports[`Pointer with children of type circle 1`] = `
<g>
  <circle
    cx={0}
    cy={10}
    r={10}
    transform="rotate(0 50 50) translate(50 -50) " />
</g>
`;

exports[`Pointer with children of type ellipse 1`] = `
<g>
  <ellipse
    cx={0}
    cy={10}
    rx={10}
    ry={10}
    transform="rotate(0 50 50) translate(50 -50) " />
</g>
`;

exports[`Pointer with children of type rect 1`] = `
<g>
  <rect
    height={10}
    transform="rotate(0 50 50) translate(50 -50) "
    width={10}
    x={-5}
    y={0} />
</g>
`;

exports[`Pointer without children use the type prop to render the pointer 1`] = `
<rect
  fill={undefined}
  filter={undefined}
  height={10}
  style={undefined}
  transform="someTransformation"
  width={10}
  x={-5}
  y={0} />
`;

exports[`Pointer without children without type prop renders a circle by default 1`] = `
<circle
  cx={0}
  cy={10}
  fill={undefined}
  filter={undefined}
  r={10}
  style={undefined}
  transform="someTransformation" />
`;
